src/index.js

```diff
@@ -1,189 +1,222 @@
+const DEFAULT_DIRECT_DOMAINS = ['*.google.com', '*.gstatic.com', '*.googleapis.com'];
+
+const requestPromised = promisify(request, Promise);
+
+function createSauceConnectProcess (options) {
+    return new Promise((resolve, reject) => {
+        sauceConnectLauncher(options, (err, process) => {
+            if (err) {
+                reject(err);
+                return;
+            }
+
+            resolve(process);
+        });
+    });
+}
+
+function disposeSauceConnectProcess (process) {
+    return new Promise(resolve => {
+        process.close(resolve);
+    });
+}
+
+export default class SaucelabsConnector {
+    constructor (username, accessKey, options = {}) {
+        const timestamp = Date.now();
+
+        const {
+            createTunnel     = true,
+            connectorLogging = true,
+        } = options;
+
+        this.options = { connectorLogging, createTunnel };
+
+        const {
+            tunnelLogging    = false,
+            tunnelIdentifier = createTunnel ? timestamp : void 0,
+            logfile          = 'sc_' + (tunnelIdentifier || timestamp) + '.log',
+            directDomains    = DEFAULT_DIRECT_DOMAINS,
+            noSSLBumpDomains = [],
+
+            ...sauceConnectOptions
+        } = options;
+
+        this.username            = username;
+        this.accessKey           = accessKey;
+        this.tunnelIdentifier    = tunnelIdentifier;
+        this.sauceConnectOptions = null;
+        this.sauceConnectProcess = null;
+
+        if (createTunnel) {
+            this.sauceConnectOptions = {
+                ...sauceConnectOptions,
+
+                username:         this.username,
+                accessKey:        this.accessKey,
+                tunnelIdentifier: this.tunnelIdentifier,
+                logfile:          tunnelLogging ? logfile : null,
+
+                ...noSSLBumpDomains.length && {
+                    noSSLBumpDomains: noSSLBumpDomains.join(',')
+                },
+
+                ...directDomains.length && {
+                    directDomains: directDomains.join(',')
+                }
+            };
+        }
+
+        this.sauceStorage = new SauceStorage(this.username, this.accessKey);
+
+        wd.configureHttp({
+            retryDelay: WEB_DRIVER_CONFIGURATION_RETRY_DELAY,
+            retries:    WEB_DRIVER_CONFIGURATION_RETRIES,
+            timeout:    WEB_DRIVER_CONFIGURATION_TIMEOUT
+        });
+
+    }
+
+    _log (message) {
+        if (this.options.connectorLogging)
+            process.stdout.write(message + '\n');
+    }
+
+    async _getFreeMachineCount () {
+        var params = {
+            method: 'GET',
+            url:    [`https://${SAUCE_API_HOST}/rest/v1/users`, this.username, 'concurrency'].join('/'),
+            auth:   { user: this.username, pass: this.accessKey }
+        };
+
+        var response = await requestPromised(params);
+
+        return JSON.parse(response.body).concurrency[this.username].remaining.overall;
```

Comment:
I propose to handle connection errors for all request and display clear messages for them.